### ISSUE TITLE
Near future directions, follow-up

### ISSUE THREAD

**Issue by fchollet** [fchollet/keras#793]
We have made good progress on our current directions, with the introduction of automated shape inference. Here's a recap to help synchronize our community efforts.
#### What should be our priority right now:
- abstracting Theano functionality into a modular Keras back-end.
#### What should be our number 2 priorities:
- Stricter checks of assumptions throughout the codebase.
- Proper docstrings everywhere as per @matsuyamax's remarks.
- As a follow-up, autogenerated docs (all classes / functions specifications).
- Simpler, faster unit tests.
- Faster compilation of complex graphs via tensor caching (see https://github.com/fchollet/keras/issues/666).
#### What we need to work on, but isn't as urgent:
- New, simpler API for the Graph model (still TBD).
- Recurrent Container (@EderSantana, you seemed to be pretty close to this).
- Stateful RNNs.
- Ability to reinitialize weights without recompiling.
- Simpler way of handling loss weighting and masking (I though @EderSantana's objective function decorator was a great idea).
- A proper parameter naming system (might be introduced as part of the backend).
- Caffe model importing (@pranv).

**Comment by EderSantana** [comment]
> Recurrent Container (@EderSantana, you seemed to be pretty close to this).

It seems to work, as long as you are willing to do everything inside the `scan` loop. I tried that common GRU optimization of precomputing input outside the loop, but there is no way to pass multiple named inputs to containers yet. Also, the new shape inference API may require me to change the logic a bit. So, progress here depends a lot on these other two choices. But be assured this is the feature I want the most and I'll work on it as soon as I see these other two parts mature enough.

> Stateful RNNs.

I have a working example of this: https://github.com/fchollet/keras/issues/763
I've also a way to of training the initial states, for cases when we don't need to go stateful. I believe that trainable initial states should be the new default. All that is needed here is to define how to change the recurrent layers. I believe we don't want a separate class for each stateful version.

> Simpler way of handling loss weighting and masking (I though @EderSantana's objective function decorator was a great idea).

Ok, so while I wait for the API choices mentioned above, I'll work on this since it seems to be independent.

**Comment by EderSantana** [comment]
> abstracting Theano functionality into a modular Keras back-end.

Just to get things started, how a `Dense` layer would look like without direct Theano reference? Is anybody already working on this?

**Comment by pranv** [comment]
@endersantana, you'd improt `tensor` or `dot` from something like `keras.backends`.

**Comment by elanmart** [comment]
> New, simpler API for the Graph model (still TBD)

@fchollet Are you sure you want to get rid of `Graph's` `add_input` method? What about changing it to `add_inputs` which woudl take a `dict` of `name:ndim` pairs? I use it in my code and IMO it's quite convenient.

**Comment by EderSantana** [comment]
@pranv That is what I imagined. Don't you think such a thing, if done well, would have uses that could benefit more projects than just Keras? Keep that in mind for the possibility of making `backends` an independent project.

**Comment by lukedeo** [comment]
@EderSantana I definitely agree with the idea of having `backends` be a separate project. I think there is a whole slew of new backend options, abstracting that into separate package would make the whole autodiff+symbolic math + GPU space a bit more manageable for new projects.

**Comment by fchollet** [comment]
> @fchollet Are you sure you want to get rid of Graph's add_input method? What about changing it to add_inputs which woudl take a dict of name:ndim pairs? I use it in my code and IMO it's quite convenient.

No, I am not sure at all, `add_input` obviously has value that isn't readily replaceable. I think being to add all inputs at once is a good suggestion. But there are actually 2 parameters we need to specify for all inputs: `input_shape` and `dtype`. So an `inputs` dict would end up being a complicated interface (mapping names to dictionaries?)

> @pranv That is what I imagined. Don't you think such a thing, if done well, would have uses that could benefit more projects than just Keras? Keep that in mind for the possibility of making backends an independent project.

Yes, basically I would expect to import the backend via: `from keras import backend as K` which would import a different set of files based on the type of backend you configured in your Keras config file (another feature we need to add). Then you would call `K.dot()`, `K.function()`, etc.

I agree that such a project could potentially become a lingua franca of Python tensor manipulation (write your projects in K, run them on any tensor framework), specifically useful for projects like Keras that don't want to be dependent on any single framework / platform [*]. However that is beyond our scope for now. Let's just implement it internally for now (with generality in mind), and if there's interest in splitting it up later we can do that. In time.

[*] About "platforms": in the future, fast hardware clouds like Nervana Systems' ASIC cloud will be available to the public (Amazon will most likely have a Deep Learning AWS service, etc). Think 100x faster than today's high-end Nvidia GPUs. But they will require using custom tensor frameworks (e.g. Neon) due to the tight coupling between low-level code and hardware. If you want to run on any platform you will need to write generic code (in K) that will compile to a given target framework (Neon).

> Ok, so while I wait for the API choices mentioned above, I'll work on this since it seems to be independent.

Awesome, let me know if you need any help. The shape inference stuff is already in master, and the Graph API changes will hopefully be settled soon.

**Comment by LeavesBreathe** [comment]
Can I just say caffe integration would be amazing
